src/ast/irbuilderbpf.cpp

```diff
@@ -450,6 +480,169 @@ Value *IRBuilderBPF::CreateMapLookupElem(Value *ctx,
+                                               const std::string &map_name,
+                                               Value *key,
+                                               Value *cpu,
+                                               SizedType &type,
+                                               const location &loc)
+{
+  assert(ctx && ctx->getType() == GET_PTR_TY());
+  CallInst *call = createPerCpuMapLookup(map_name, key, cpu);
+
+  // Check if result == 0
+  Function *parent = GetInsertBlock()->getParent();
+  BasicBlock *lookup_success_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_success",
+                                                        parent);
+  BasicBlock *lookup_failure_block = BasicBlock::Create(module_.getContext(),
+                                                        "lookup_failure",
+                                                        parent);
+  BasicBlock *lookup_merge_block = BasicBlock::Create(module_.getContext(),
+                                                      "lookup_merge",
+                                                      parent);
+
+  AllocaInst *value = CreateAllocaBPF(type, "lookup_elem_val");
+  Value *condition = CreateICmpNE(
+      CreateIntCast(call, GET_PTR_TY(), true),
+      ConstantExpr::getCast(Instruction::IntToPtr, getInt64(0), GET_PTR_TY()),
+      "map_lookup_cond");
+  CreateCondBr(condition, lookup_success_block, lookup_failure_block);
+
+  SetInsertPoint(lookup_success_block);
+  // per cpu maps used by bpftrace only hold integer types for aggregation
+  // so we don't need to memcpy
+  assert(value->getAllocatedType() == getInt64Ty());
+  // createMapLookup  returns an u8*
+  auto *cast = CreatePointerCast(call, value->getType(), "cast");
+  CreateStore(CreateLoad(getInt64Ty(), cast), value);
+  CreateBr(lookup_merge_block);
+
+  SetInsertPoint(lookup_failure_block);
+
+  CreateStore(getInt64(0), value);
+  CreateHelperError(
+      ctx, getInt32(0), libbpf::BPF_FUNC_map_lookup_percpu_elem, loc);
+  CreateBr(lookup_merge_block);
+
+  SetInsertPoint(lookup_merge_block);
+
+  // value is a pointer to i64
+  Value *ret = CreateLoad(getInt64Ty(), value);
+  CreateLifetimeEnd(value);
+  return ret;
+}
+
+Value *IRBuilderBPF::CreatePerCpuMapSumElems(Value *ctx,
+                                             Map &map,
+                                             Value *key,
+                                             const location &loc,
+                                             bool is_aot)
+{
+  assert(ctx && ctx->getType() == GET_PTR_TY());
+
+  const std::string &map_name = map.ident;
+
+  AllocaInst *sum = CreateAllocaBPF(getInt64Ty(), "sum");
+  AllocaInst *i = CreateAllocaBPF(getInt32Ty(), "i");
+
+  // Set a large upper bound if we don't know the number of cpus
+  // when generating the instructions
+  int nr_cpus = is_aot ? 1024 : bpftrace_.get_num_possible_cpus();
```

Comment:
I forget if we discussed this in person, did it not work to pass the nr_cpu from userspace and then mask (`&=`) the count?

If it didn't and we have to use this method, since 1024 is a fixed value, can bpftrace check at runtime if there's > 1024 CPUs and warn about it? This is a potential for incorrect results & and we should be really careful in these situations. Should be possible in a shared codepath between AOT and dynamic mode.

Reply:
> I forget if we discussed this in person, did it not work to pass the nr_cpu from userspace and then mask (&=) the count?

I've tried other limiting techniques in other PRs and it seems like the verifier is unhappy if it doesn't know the value in advance and the type can exceed the max number of loops. You can normally get around this using libbpf by having a mutable volatile with an initial value and modifying it using the skeleton in userspace but we don't have that yet. @ajor and I discussed the option of doing this by modifying the rodata section but decided just to punt on that since AOT isn't officially supported yet anyway. 

I will add a warning if we detect the number of cpus to be greater than 1024 in userspace.

Reply:
I also have plans to add proper loops using open coded iterators which should also fix this issue.